File: fend/uek223-gruppe2-frontend/react_frontend/src/Services/BlogPostService.ts
import api from '../config/Api';
import { BlogPost, BlogPostPageResponse } from '../types/models/BlogPost.model';

/**
 * Query parameters for fetching blog posts with pagination, sorting, and filtering
 */
export interface BlogPostQueryParams {
  page?: number;
  size?: number;
  sortBy?: 'date' | 'title' | 'category';
  sortOrder?: 'asc' | 'desc';
  category?: string;
  authorId?: string;
}

/**
 * BlogPostService - API service for blog post CRUD operations
 * Handles all HTTP requests related to blog posts.
 */
const BlogPostService = {

  /**
   * Creates a new blog post (UC1)
   */
  createBlogPost: (blogPost: BlogPost) => {
    return api.post('/blogpost', blogPost);
  },

  /**
   * Fetches all blog posts with optional pagination, sorting, and filtering (UC4)
   */
  getAllBlogPosts: (params?: BlogPostQueryParams) => {
    const queryParams = new URLSearchParams();

    if (params) {
      if (params.page !== undefined) queryParams.append('page', params.page.toString());
      if (params.size !== undefined) queryParams.append('size', params.size.toString());
      if (params.sortBy) queryParams.append('sortBy', params.sortBy);
      if (params.sortOrder) queryParams.append('sortOrder', params.sortOrder);
      if (params.category) queryParams.append('category', params.category);
      if (params.authorId) queryParams.append('authorId', params.authorId);
    }

    const queryString = queryParams.toString();
    return api.get<BlogPostPageResponse>(`/blogpost${queryString ? `?${queryString}` : ''}`);
  },

  /**
   * Fetches a single blog post by ID (UC4)
   */
  getBlogPostById: (id: string) => {
    return api.get(`/blogpost/${id}`);
  },

  /**
   * Updates an existing blog post (UC2, UC5.1)
   */
  updateBlogPost: (id: string, blogPost: BlogPost) => {
    return api.put(`/blogpost/${id}`, blogPost);
  },

  /**
   * Deletes a blog post by ID (UC3, UC5.2)
   */
  deleteBlogPost: (id: string) => {
    return api.delete(`/blogpost/${id}`);
  }
};

export default BlogPostService;

File: bend/uek223-gruppe2-backend/spring_backend/src/main/java/com/example/demo/domain/blogpost/BlogPostService.java
package com.example.demo.domain.blogpost;

import com.example.demo.core.generic.AbstractService;
import com.example.demo.domain.blogpost.dto.BlogPostUpdateDTO;
import com.example.demo.domain.user.User;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

public interface BlogPostService extends AbstractService<BlogPost> {

  BlogPost createBlogPost(BlogPost blogPost, User author);

  /**
   * UC2/UC5: Update blog post with validation
   * 
   * @param id        Blog post ID to update
   * @param updateDTO Blog post update data
   * @return Updated blog post
   */
  BlogPost updateBlogPost(UUID id, BlogPostUpdateDTO updateDTO);

  /**
   * UC4: Find all blog posts with pagination, sorting, and optional filtering
   * 
   * @param pageable Pagination and sorting parameters
   * @param category Optional category filter
   * @param authorId Optional author ID filter
   * @return Paginated blog posts
   */
  Page<BlogPost> findAllWithFilters(Pageable pageable, String category, UUID authorId);

}
